### ISSUE TITLE
Failure to disconnect after server disappears

### ISSUE THREAD

**Issue by Gilead** [xmppjs/xmpp.js#562]
Steps to reproduce:
0. Have a simple xmpp.js client running (nothing fancy -- basic auth and callbacks)
1. XMPP server disappears (for example, killed forcibly without a chance to close the connection properly)
2. xmpp.js emits status 'disconnect'
3. execute the following code in an attempt to clear the connection status before reconnecting (attempting to call client.start(address) straight away fails because 'connection is not offline')
```
        function () {
            return client.stop().
                catch(function (e) {
                console.log("XMPP failed to stop because: %o", e);
                return undefined;
            }).then(client.disconnect).
                catch(function (e) {
                console.log("XMPP failed to disconnect because: %o", e);
                return undefined;
            });
        };
```
4. observe failure because
```
TypeError: Cannot read property 'timeout' of undefined
    at disconnect (/node_modules/@xmpp/connection/index.js:201:24)
```
(line 201 is ```disconnect(ms = this.timeout) {```)
5. modify the code to provide a timeout parameter
```
        function () {
            return client.stop().
                catch(function (e) {
                console.log("XMPP failed to stop because: %o", e);
                return undefined;
            }).then(function () { return client.disconnect(2000); }).
                catch(function (e) {
                console.log("XMPP failed to disconnect because: %o", e);
                return undefined;
            });
        };
```
6. observe failure because
``` 
TypeError: Cannot read property 'end' of null
    at Client.disconnect (/node_modules/@xmpp/connection/index.js:203:17)
```
(line 203 is ```this.socket.end()```)

It appears something is not right with the 'this' context.

This is an issue because I would prefer my xmpp.js client to NOT have to wait several minutes until, presumably, connection times out, before it attempts to reconnect.

**Comment by sonnyp** [comment]
@Gilead I'm pretty sure this was fixed in master

Could you try with master and let us know?

**Comment by yuanchieh-cheng** [comment]
I encounter the same problem here.  
When I manually close xmpp server the client will throw error without retry.
```
{ Error: read ECONNRESET
    at _errnoException (util.js:992:11)
    at TCP.onread (net.js:618:25) code: 'ECONNRESET', errno: 'ECONNRESET', syscall: 'read' }
unhandledRejection { Error: read ECONNRESET
    at _errnoException (util.js:992:11)
    at TCP.onread (net.js:618:25) code: 'ECONNRESET', errno: 'ECONNRESET', syscall: 'read' }
send to slack
```

using 0.3.0

**Comment by sonnyp** [comment]
fixed in 0.5.0 unfortunally there is a lot of API breaking changes, please refer to the documentation https://github.com/xmppjs/xmpp.js/tree/master/packages/client
